### ISSUE TITLE
Shell app fails to open

### ISSUE THREAD

**Issue by nobodyme** [OSC/ondemand#2369]
Shell apps fails to open and I'm greeted with an error page, reading,

```
We are sorry something went wrong. This issue has been logged for investigation. Please try again later
```

I see this logged in on `/var/log/ondemand-nginx/<user>/error.log`

```
App 17221 output: innerError Error: Cannot find module '../build/Debug/pty.node'
   81 App 17221 output: Require stack:
   82 App 17221 output: - /var/www/ood/apps/sys/shell/node_modules/node-pty/lib/unixTerminal.js
   83 App 17221 output: - /var/www/ood/apps/sys/shell/node_modules/node-pty/lib/index.js
   84 App 17221 output: - /var/www/ood/apps/sys/shell/app.js
   85 App 17221 output: - /opt/ood/ondemand/root/usr/share/passenger/helper-scripts/node-loader.js
   86 App 17221 output:   at Function.Module._resolveFilename (internal/modules/cjs/loader.js:902:15)
   87 App 17221 output:   at Function.Module._load (internal/modules/cjs/loader.js:746:27)
   88 App 17221 output:   at Module.require (internal/modules/cjs/loader.js:974:19)
   89 App 17221 output:   at Module.require (/opt/ood/ondemand/root/usr/share/passenger/helper-scripts/node-loader.js:80:25)
   90 App 17221 output:   at require (internal/modules/cjs/helpers.js:101:18)
   91 App 17221 output:   at Object.<anonymous> (/var/www/ood/apps/sys/shell/node_modules/node-pty/lib/unixTerminal.js:30:15)
   92 App 17221 output:   at Module._compile (internal/modules/cjs/loader.js:1085:14)
   93 App 17221 output:   at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
   94 App 17221 output:   at Module.load (internal/modules/cjs/loader.js:950:32)
   95 App 17221 output:   at Function.Module._load (internal/modules/cjs/loader.js:790:12) {
   96 App 17221 output:  code: 'MODULE_NOT_FOUND',
   97 App 17221 output:  requireStack: [
   98 App 17221 output:   '/var/www/ood/apps/sys/shell/node_modules/node-pty/lib/unixTerminal.js',
   99 App 17221 output:   '/var/www/ood/apps/sys/shell/node_modules/node-pty/lib/index.js',
  100 App 17221 output:   '/var/www/ood/apps/sys/shell/app.js',
  101 App 17221 output:   '/opt/ood/ondemand/root/usr/share/passenger/helper-scripts/node-loader.js'
  102 App 17221 output:  ]
  103 App 17221 output: }
  104 App 17221 output: /var/www/ood/apps/sys/shell/node_modules/node-pty/lib/unixTerminal.js:35
  105 App 17221 output:     throw outerError;
  106 App 17221 output:     ^
  107 App 17221 output:
  108 App 17221 output: Error: The module '/var/www/ood/apps/sys/shell/node_modules/node-pty/build/Release/pty.node'
  109 App 17221 output: was compiled against a different Node.js version using
  110 App 17221 output: NODE_MODULE_VERSION 72. This version of Node.js requires
  111 App 17221 output: NODE_MODULE_VERSION 83. Please try re-compiling or re-installing
  112 App 17221 output: the module (for instance, using `npm rebuild` or `npm install`).
  113 App 17221 output:   at Object.Module._extensions..node (internal/modules/cjs/loader.js:1144:18)
  114 App 17221 output:   at Module.load (internal/modules/cjs/loader.js:950:32)
  115 App 17221 output:   at Function.Module._load (internal/modules/cjs/loader.js:790:12)
  116 App 17221 output:   at Module.require (internal/modules/cjs/loader.js:974:19)
  117 App 17221 output:   at Module.require (/opt/ood/ondemand/root/usr/share/passenger/helper-scripts/node-loader.js:80:25)
  118 App 17221 output:   at require (internal/modules/cjs/helpers.js:101:18)
  119 App 17221 output:   at Object.<anonymous> (/var/www/ood/apps/sys/shell/node_modules/node-pty/lib/unixTerminal.js:26:11)
  120 App 17221 output:   at Module._compile (internal/modules/cjs/loader.js:1085:14)
  121 App 17221 output:   at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
  122 App 17221 output:   at Module.load (internal/modules/cjs/loader.js:950:32) {
  123 App 17221 output:  code: 'ERR_DLOPEN_FAILED'
  124 App 17221 output: }
  125 [ E 2022-11-17 10:27:49.9685 17200/Tf age/Cor/App/Implementation.cpp:221 ]: Could not spawn process for application /var/www/ood/apps/sys/shell: The application process exited prematurely.
  126  Error ID: 12e67ceb
  127  Error details saved to: /tmp/passenger-error-nzvrmQ.html
```

Please let me know, if any other details are required. Thanks

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1202821104799596/1203392141682048) by [Unito](https://www.unito.io)

**Comment by nobodyme** [comment]
Thanks for the quick response,
Platform - CentOS 7
Ondemand - 2.0.18

**Comment by johrstrom** [comment]
OK sorry for the confusion.  `2.0.18` needs node 12 (NODE_MODULE_VERSION 72).  `2.0.29` got upgraded to node 14 (NODE_MODULE_VERSION 83) because node 12 is end of life (not receiving security updates anymore).

Since you're on centos 7, you'll need `rh-nodejs12` if you want to keep 2.0.18 and you seem to have `rh-nodejs14`.

Again - sorry for the disruption. node going EOL kinda forced our hands.

**Comment by treydock** [comment]
@nobodyme What were the steps taken to install OnDemand? If you installed via the YUM repo the dependencies should have been pulled in automatically.

**Comment by nobodyme** [comment]
Thanks @johrstrom, for now I upgraded to `2.0.29` and I don't see the issue.

Hey @treydock,
It does pull in the dependencies automatically, so unless `2.0.18` had suddenly changed it's dependency from 12 to 14, I shouldn't be seeing an issue. Investigating further on this.

**Comment by nobodyme** [comment]
Confirmed this is what is happening, `2.0.18` does seem to pull in `rh-nodejs14`. If I understood, @johrstrom correctly, this shouldn't happen right?

**Comment by treydock** [comment]
It depends on how things were installed.  If you installed OnDemand 2.0.18 after we published 2.0.29 then what you got is unfortunately expected behavior.  The 2.0.18 release allowed the meta packages that pulled in nodejs14:

https://github.com/OSC/ondemand/blob/566662105b789b19e057b669e9b9c13bd6182490/packaging/ondemand.spec#L71

Future releases of OnDemand 2.0 won't have this issue as the RPM dependency is more specific: 

https://github.com/OSC/ondemand/blob/a7d3ee8ffbaa15c8534e01199a7a8ee6d67c7436/packaging/rpm/ondemand.spec#L67

**Comment by nobodyme** [comment]
Ohh, that's unfortunate, yes I did install after 2.0.29 was released, thanks for the confirmation.
